File: 01.linked list/python/linked_list.py
class Node:
    def __init__(self, v):
        self.value = v
        self.next = None


class LinkedList:

    def __init__(self):
        self.head = None
        self.tail = None

    def add_in_tail(self, item):
        if self.head is None:
            self.head = item
        else:
            self.tail.next = item
        self.tail = item

    def print_all_nodes(self):
        node = self.head
        while node != None:
            print(node.value)
            node = node.next

    def find(self, val):
        node = self.head
        while node is not None:
            if node.value == val:
                return node
            node = node.next
        return None

    def find_all(self, val):
        nodes = []
        node = self.head
        while node is not None:
            if node.value == val:
                nodes += [node]
            node = node.next
        return nodes

    def delete(self, val, all=False):
        if self.head is None:  # Check if list is not empty
            return

        node = self.head
        while node is not None and node.value == val:  # Delete first node(s) that are equal(s) given value
            self.head = self.head.next
            node = self.head
            if self.head is None:
                self.tail = None
            if not all:
                return

        previous_node = None
        while node is not None:  # Delete node(s) in middle
            while node is not None and node.value != val:
                previous_node = node
                node = node.next

            if node is None:
                return

            previous_node.next = node.next
            node = previous_node.next
            if previous_node.next is None:
                self.tail = previous_node

            if not all:
                break

    def clean(self):
        node = self.head
        pointer = node.next
        while pointer is not None:
            node.next = pointer.next
            node = pointer
            pointer = pointer.next
        self.head = None
        self.tail = None

    def len(self):
        node = self.head
        count = 0
        while node is not None:
            count += 1
            node = node.next
        return count

    def insert(self, afterNode, newNode):
        if self.head is None:
            self.add_in_tail(newNode)
        node = self.head
        while node is not None:
            if node == afterNode:
                newNode.next = node.next
                node.next = newNode
                if node.next.next is None:
                    self.tail = node.next
                break
            node = node.next
        return

File: 01.linked list/golang/linked_list.go
package main

import (
	"errors"
)

type Node struct {
	next  *Node
	value int
}

type LinkedList struct {
	head *Node
	tail *Node
}

func (l *LinkedList) AddInTail(item Node) {
	if l.head == nil {
		l.head = &item
	} else {
		l.tail.next = &item
	}

	l.tail = &item
}

func (l *LinkedList) Find(n int) (Node, error) {
	node := l.head
	for node != nil {
		if node.value == n {
			return *node, nil
		}
		node = node.next
	}
	return Node{value: -1, next: nil}, errors.New("couldn't find node with such value")
}

func (l *LinkedList) FindAll(n int) []Node {
	nodes := make([]Node, 0)
	node := l.head
	for node != nil {
		if node.value == n {
			// append a detached copy (next = nil) for predictable equality in tests
			nodes = append(nodes, Node{value: node.value})
		}
		node = node.next
	}
	return nodes
}

func (l *LinkedList) Delete(n int, all bool) {
	if l.head == nil {
		return
	}

	node := l.head

	// Delete from the beginning while head matches
	for node != nil && node.value == n {
		l.head = l.head.next
		node = l.head
		if l.head == nil {
			l.tail = nil
		}
		if !all {
			return
		}
	}

	var prev *Node
	for node != nil {
		for node != nil && node.value != n {
			prev = node
			node = node.next
		}

		if node == nil {
			return
		}

		// remove node
		prev.next = node.next
		node = prev.next
		if prev.next == nil {
			l.tail = prev
		}

		if !all {
			break
		}
	}
}

func (l *LinkedList) Count() int {
	count := 0
	node := l.head
	for node != nil {
		count++
		node = node.next
	}
	return count
}

func (l *LinkedList) Insert(after *Node, add Node) {
	if after == l.tail {
		l.AddInTail(add)
		return
	}

	if after == nil {
		add.next = l.head
		l.head = &add
		if l.tail == nil {
			l.tail = &add
		}
		return
	}

	node := l.head
	for node != nil {
		if node == after {
			add.next = node.next
			node.next = &add
			if node.next != nil && node.next.next == nil {
				l.tail = node.next
			}
			break
		}
		node = node.next
	}
}

func (l *LinkedList) InsertFirst(first Node) {
	if l.head == nil {
		l.head = &first
		l.tail = &first
		return
	}
	first.next = l.head
	l.head = &first
}

func (l *LinkedList) Clean() {
	l.head = nil
	l.tail = nil
}
